### PR TITLE
feat: `esl-toggleable`, `esl-panel`, `esl-alert` migration to `esl-event-listener`

### PR DESCRIPTION
**Pull request by ala-n**

### CODE REVIEW

**Review by yadamskaya — APPROVED** [review]

**Review by NastaLeo — APPROVED** [review]

File: src/modules/esl-toggleable/core/esl-toggleable.ts [hunk 1]
```diff
@@ -56,6 +56,29 @@ export class ESLToggleable extends ESLBaseElement {
...
+  @prop('esl:show') public SHOW_EVENT: string;
+  /** Event to dispatch when toggleable is deactivated */
+  @prop('esl:hide') public HIDE_EVENT: string;
+
+  /** Event to dispatch when toggleable has end activation process */
+  @prop('esl:after:show') public AFTER_SHOW_EVENT: string;
+  /** Event to dispatch when toggleable has end deactivation process */
+  @prop('esl:after:hide') public AFTER_HIDE_EVENT: string;
+
+  /** Event to activate toggleables on event way */
```
Comment (dshovchko) [inline src/modules/esl-toggleable/core/esl-toggleable.ts:74]:
```suggestion
  /** Event to activate toggleables by event way */
```

File: src/modules/esl-toggleable/core/esl-toggleable.ts [hunk 2]
```diff
@@ -56,6 +56,29 @@ export class ESLToggleable extends ESLBaseElement {
...
+  @prop('esl:show:request') public SHOW_REQUEST_EVENT: string;
+  /** Event to deactivate toggleables on event way */
```
Comment (dshovchko) [inline src/modules/esl-toggleable/core/esl-toggleable.ts:76]:
```suggestion
  /** Event to deactivate toggleables by event way */
```

File: src/modules/esl-toggleable/core/esl-toggleable.ts [hunk 3]
```diff
@@ -167,12 +167,8 @@ export class ESLToggleable extends ESLBaseElement {
...
     this._trackHover = track;
 
-    this.removeEventListener('mouseenter', this._onMouseEnter);
-    this.removeEventListener('mouseleave', this._onMouseLeave);
-    if (this._trackHover) {
-      this.addEventListener('mouseenter', this._onMouseEnter);
-      this.addEventListener('mouseleave', this._onMouseLeave);
-    }
+    this._trackHover ? this.$$on(this._onMouseEnter) : this.$$off(this._onMouseEnter);
+    this._trackHover ? this.$$on(this._onMouseLeave) : this.$$off(this._onMouseLeave);
```
Comment (dshovchko) [inline src/modules/esl-toggleable/core/esl-toggleable.ts:171]:
```suggestion
    track ? this.$$on(this._onMouseEnter) : this.$$off(this._onMouseEnter);
    track ? this.$$on(this._onMouseLeave) : this.$$off(this._onMouseLeave);
```
such declaration is more compact

File: src/modules/esl-toggleable/core/esl-toggleable.ts [hunk 4]
```diff
@@ -285,50 +281,71 @@ export class ESLToggleable extends ESLBaseElement {
...
+  protected buildRequestParams(e: CustomEvent<ESLToggleableRequestDetails>): ToggleableActionParams | null {
+    const detail = e.detail || {};
+    if (!isMatches(this, detail.match)) return null;
+    const params: ToggleableActionParams = Object.assign({}, detail.params || {}, {event: e});
+    if (e.type === this.SHOW_REQUEST_EVENT && typeof detail.delay === 'number') {
+      Object.assign(params, {showDelay: detail.delay});
+    }
+    if (e.type === this.HIDE_REQUEST_EVENT && typeof detail.delay === 'number') {
+      Object.assign(params, {showDelay: detail.delay});
+    }
```
Comment (dshovchko) [inline src/modules/esl-toggleable/core/esl-toggleable.ts:334]:
```suggestion
    if ((e.type === this.SHOW_REQUEST_EVENT || e.type === this.HIDE_REQUEST_EVENT) && typeof detail.delay === 'number') {
      Object.assign(params, {showDelay: detail.delay});
    }
```

**Review by dshovchko — APPROVED** [review]

**Review by dshovchko — APPROVED** [review]

File: src/modules/esl-toggleable/README.md [hunk 5]
```diff
@@ -9,7 +9,7 @@ Authors: *Julia Murashko*, *Alexey Stsefanovich (ala'n)*
 ESLToggleable - a custom element, that is used as a base for "Popup-like" components creation.
 
 ESLToggleableDispatcher - plugin component, that prevents activation of multiple ESLToggleable instances in bounds of managed container.
-Usually (and by default) binds to document.body. Use native DOM events to manage controlled instances state.
+Usually (and by default) binds to `document.body`. Use native DOM events to manage controlled instances state.
```
Comment (NastaLeo) [inline src/modules/esl-toggleable/README.md:12]:
```suggestion
Usually (and by default) binds to `document.body`. Uses native DOM events to manage controlled instances state.
```

File: src/modules/esl-toggleable/core/esl-toggleable.ts [hunk 6]
```diff
@@ -285,50 +280,64 @@ export class ESLToggleable extends ESLBaseElement {
...
-  @bind
+  @listen({auto: false, event: 'mouseleave'})
   protected _onMouseLeave(e: MouseEvent): void {
     const hideDelay = this._trackHoverDelay;
-    const baseParams: ToggleableActionParams = {initiator: 'mouseleave', trackHover: true, activator: this.activator, event: e, hideDelay};
+    const baseParams: ESLToggleableActionParams = {initiator: 'mouseleave', trackHover: true, activator: this.activator, event: e, hideDelay};
     this.hide(Object.assign(baseParams, this.trackHoverParams));
   }
 
+  /** Prepare toggle request events param */
```
Comment (NastaLeo) [inline src/modules/esl-toggleable/core/esl-toggleable.ts:323]:
```suggestion
  /** Prepares toggle request events param */
```

File: src/modules/esl-alert/core/esl-alert.ts [hunk 7]
```diff
@@ -140,16 +128,25 @@ export class ESLAlert extends ESLToggleable {
...
-      this.hide(params);
+    if (e.type === this.HIDE_REQUEST_EVENT) {
+      return Object.assign({hideDelay: 0}, e.detail, {force: true});
     }
+    return null;
+  }
+
+  @listen({inherit: true, target: (el: ESLAlert) => el.$target})
+  protected override _onHideRequest(e: CustomEvent<ESLToggleableRequestDetails>): void {
+    super._onShowRequest(e);
```
Comment (yadamskaya) [inline src/modules/esl-alert/core/esl-alert.ts:143]: It seems to me `super._onHideRequest(e);` should be here

Comment (NastaLeo) [inline src/modules/esl-alert/core/esl-alert.ts:143]:
```suggestion
    super._onHideRequest(e);
```

**Review by NastaLeo — APPROVED** [review]

**Review by abarmina — APPROVED** [review]

File: src/modules/esl-toggleable/test/toggleable.request.test.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
+import {ESLToggleable} from '../core/esl-toggleable';
+import {EventUtils} from '../../esl-event-listener/core/api';
+import type {ESLToggleableRequestDetails} from '../core/esl-toggleable';
+
+describe('ESLToggleable show/hide request events', () => {
```
[the remark this answers is not in the document]
Comment (ala-n) [reply src/modules/esl-toggleable/test/toggleable.request.test.ts:5]: I think no for that one, at least

**Pull request merged**